sql/core/src/main/scala/org/apache/spark/sql/RelationalGroupedDataset.scala

```diff
@@ -529,7 +530,31 @@ class RelationalGroupedDataset protected[sql](
         initialStateSchema = initialState.df.schema
       )
     }
-    Dataset.ofRows(df.sparkSession, plan)
+
+    if (eventTimeColumnName.isEmpty) {
+      Dataset.ofRows(df.sparkSession, plan)
+    } else {
+      updateEventTimeColumnAfterTransformWithState(plan, eventTimeColumnName)
+    }
+  }
+
+  /**
+   * Creates a new dataset with updated eventTimeColumn after the transformWithState
+   * logical node.
+   */
+  private def updateEventTimeColumnAfterTransformWithState(
+      transformWithStateInPandas: LogicalPlan,
+      eventTimeColumnName: String): DataFrame = {
+    val transformWithStateDataset = Dataset.ofRows(
+      df.sparkSession,
+      transformWithStateInPandas
+    )
+
+    Dataset.ofRows(df.sparkSession, EliminateEventTimeWatermark(
```

Comment:
EliminateEventTimeWatermark is no longer needed as we now have this in Analyzer rule.

But maybe we could deal with this in separate PR to address other cases as well.

Reply:
> EliminateEventTimeWatermark is no longer needed as we now have this in Analyzer rule.

Is this already done? I saw we still keep the TODO here: https://github.com/apache/spark/blob/d6b733492d3e1643babbcd2b1860e35af7928906/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala#L3933

Reply:
Yeah I forgot to remove the TODO. My bad. It's added.